pxtlib/auth.ts

```diff
@@ -84,29 +84,70 @@ namespace pxt.auth {
     let debouncePreferencesChangedTimeout = 0;
     let debouncePreferencesChangedStarted = 0;
 
-    export abstract class AuthClient {
-        /**
-         * In-memory cache of user profile state. This is persisted in local storage.
-         * DO NOT ACCESS DIRECTLY UNLESS YOU KNOW IT IS OK. Functions allowed direct
-         * access are marked as such below.
-         */
-        private state$: Readonly<State>;
+    //
+    // Local storage of auth token.
+    //
+
+    // Last known auth token state. This is provided as a convenience for legacy methods that cannot be made async.
+    // Preference hasAuthTokenAsync() over taking a dependency on this cached value.
+    export let cachedHasAuthToken = false;
+
+    export async function getAuthTokenAsync(): Promise<string> {
+        let token: string;
+        try { token = await pxt.storage.shared.getAsync(AUTH_CONTAINER, CSRF_TOKEN_KEY); } catch {}
+        cachedHasAuthToken = !!token;
+        return token;
+    }
+    async function setAuthTokenAsync(token: string): Promise<void> {
+        cachedHasAuthToken = !!token;
+        return await pxt.storage.shared.setAsync(AUTH_CONTAINER, CSRF_TOKEN_KEY, token);
+    }
+    export async function hasAuthTokenAsync(): Promise<boolean> {
+        return !!(await getAuthTokenAsync());
```

Comment:
Is the function call necessary? Could we just use `cachedHasAuthToken` or does it not get set yet? Or is it that we don't want to look at the cached token here?

Reply:
See comment above `cachedHasAuthToken`. `hasAuthTokenAsync` is the reliable way to get this value. `cachedHasAuthToken` may not be correct, or not initialized at all. It is there for legacy compatibility only.